src/main/drivers/mcu/apm32/timer_apm32.c

```diff
@@ -0,0 +1,1198 @@
+
+/*
+    Groups that allow running different period (ex 50Hz servos + 400Hz throttle + etc):
+    TIM1 2 channels
+    TIM2 4 channels
+    TIM3 4 channels
+    TIM4 4 channels
+*/
+
+/// TODO: DAL in a lot af calls lookupTimerIndex is used. Instead of passing the timer instance the index should be passed.
+#define USED_TIMER_COUNT BITCOUNT(USED_TIMERS)
+
+#define TIM_IT_CCx(ch) (TMR_IT_CC1 << ((ch) / 4))
+
+typedef struct timerConfig_s {
+    // per-timer
+    timerOvrHandlerRec_t *updateCallback;
+
+    // per-channel
+    timerCCHandlerRec_t *edgeCallback[CC_CHANNELS_PER_TIMER];
+    timerOvrHandlerRec_t *overflowCallback[CC_CHANNELS_PER_TIMER];
+
+    // state
+    timerOvrHandlerRec_t *overflowCallbackActive; // null-terminated linkded list of active overflow callbacks
+    uint32_t forcedOverflowTimerValue;
+} timerConfig_t;
+timerConfig_t timerConfig[USED_TIMER_COUNT + 1];
+
+typedef struct {
+    channelType_t type;
+} timerChannelInfo_t;
+timerChannelInfo_t timerChannelInfo[TIMER_CHANNEL_COUNT];
+
+typedef struct {
+    uint8_t priority;
+} timerInfo_t;
+timerInfo_t timerInfo[USED_TIMER_COUNT + 1];
+
+typedef struct {
+    TMR_HandleTypeDef Handle;
+} timerHandle_t;
+timerHandle_t timerHandle[USED_TIMER_COUNT + 1];
+
+// return index of timer in timer table. Lowest timer has index 0
+#define TIMER_INDEX(i) BITCOUNT((TIM_N(i) - 1) & USED_TIMERS)
+
+static uint8_t lookupTimerIndex(const TMR_TypeDef *tim)
+{
+#define _CASE_SHF 10           // amount we can safely shift timer address to the right. gcc will throw error if some timers overlap
+#define _CASE_(tim, index) case ((unsigned)tim >> _CASE_SHF): return index; break
+#define _CASE(i) _CASE_(TMR##i##_BASE, TIMER_INDEX(i))
+
+// let gcc do the work, switch should be quite optimized
+    switch ((unsigned)tim >> _CASE_SHF) {
+#if USED_TIMERS & TIM_N(1)
+        _CASE(1);
+#endif
+#if USED_TIMERS & TIM_N(2)
+        _CASE(2);
+#endif
+#if USED_TIMERS & TIM_N(3)
+        _CASE(3);
+#endif
+#if USED_TIMERS & TIM_N(4)
+        _CASE(4);
+#endif
+#if USED_TIMERS & TIM_N(5)
+        _CASE(5);
+#endif
+#if USED_TIMERS & TIM_N(6)
+        _CASE(6);
+#endif
+#if USED_TIMERS & TIM_N(7)
+        _CASE(7);
+#endif
+#if USED_TIMERS & TIM_N(8)
+        _CASE(8);
+#endif
+#if USED_TIMERS & TIM_N(9)
+        _CASE(9);
+#endif
+#if USED_TIMERS & TIM_N(10)
+        _CASE(10);
+#endif
+#if USED_TIMERS & TIM_N(11)
+        _CASE(11);
+#endif
+#if USED_TIMERS & TIM_N(12)
+        _CASE(12);
+#endif
+#if USED_TIMERS & TIM_N(13)
+        _CASE(13);
+#endif
+#if USED_TIMERS & TIM_N(14)
+        _CASE(14);
+#endif
+#if USED_TIMERS & TIM_N(15)
+        _CASE(15);
+#endif
+#if USED_TIMERS & TIM_N(16)
+        _CASE(16);
+#endif
+#if USED_TIMERS & TIM_N(17)
+        _CASE(17);
+#endif
+#if USED_TIMERS & TIM_N(20)
+        _CASE(20);
+#endif
+    default:  return ~1;  // make sure final index is out of range
+    }
+#undef _CASE
+#undef _CASE_
+}
+
+TIM_TypeDef * const usedTimers[USED_TIMER_COUNT] = {
```

Comment:
It would be better to special-case _DEF for APM32 and keep single copy of USED_TIMERS enumeration in BF code